# SENG 474
# Assignment 1 - Problem 7
# Nolan Kurylo
# V00893175
# References:

1) https://pandas.pydata.org/pandas-docs/stable/reference/index.html

2) https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html





In [1]:

import pandas as pd
import numpy as np
import pprint
from matplotlib import pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn import tree
from sklearn.metrics import accuracy_score, zero_one_loss
from sklearn.ensemble import RandomForestClassifier

np.random.seed(1337)

df = pd.read_csv('elections_clean.csv') # create dataframe of csv


label_vector = df.pop('Democrat') # this is the target/label vector

feature_list = ['Education','Religion','EthnicMale','EthnicFemale'] # categorical features 

df = df[feature_list] # only split on categorical features (same as question 2)
df = pd.get_dummies(df, feature_list) # one hot encoding of categorical features
df['Democrat'] = label_vector # place label vector at the end of df


training_set, validation_set, training_label, validation_label = train_test_split(df.drop('Democrat', axis=1), df['Democrat'], train_size=0.7) # split 70% traing, 30% validation
RandomForest = RandomForestClassifier(n_estimators=50, max_depth=3, criterion='entropy') # generate random forest 
RandomForest.fit(training_set, training_label) # train the forest using the 70% training set

training_prediction = RandomForest.predict(training_set) #predict accuracy of tree using training set
validation_prediction = RandomForest.predict(validation_set) #predict accuracy of tree using validation set

training_acc = accuracy_score(training_label, training_prediction) * 100 # training set accuracy
training_err = zero_one_loss(training_label, training_prediction) * 100 # training set error

validation_acc = accuracy_score(validation_label, validation_prediction) *100 # training set accuracy
validation_err = zero_one_loss(validation_label, validation_prediction) * 100 # training set error


print("The Training accuracy of the RandomForest: " + str(training_acc) + "%")
print("The Training error of the RandomForest: " + str(training_err) + "%")
print()

print("The Validation accuracy of the RandomForest: " + str(validation_acc) + "%")
print("The Validation error of the RandomForest: " + str(validation_err) + "%")





The Training accuracy of the RandomForest: 87.8691503861881%
The Training error of the RandomForest: 12.1308496138119%

The Validation accuracy of the RandomForest: 88.34745762711864%
The Validation error of the RandomForest: 11.652542372881358%


# Discussion of Random Forests
The random forests actually end up having slightly worse results than singular Decision Trees. 50 decision trees in the forest pre-pruned to a depth of 3 produces a validation prediction accuracy of ~88% whereas a single Decision Tree (such as from question 2 or 4) has a prediction accuracy of ~91%. Random Forests have the purpose of increasing the accuracy of singular decision trees when the dataset the decision tree is being run on is not large enough to lead to a sufficient prediction. In the case of this large elections_clean dataset, there is enough data for a single decision tree to obtain a high level of prediction accuracy. In this case, the random forest does slightly worse as it is taking an average of numerous decision trees, inturn, being less accurate.